src/libraries/VanityAddressLib.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+/// @title VanityAddressLib
+/// @notice A library to score addresses based on their vanity
+library VanityAddressLib {
+    /// @notice Compares two addresses and returns true if the first address has a better vanity score
+    /// @param first The first address to compare
+    /// @param second The second address to compare
+    /// @return better True if the first address has a better vanity score
+    function betterThan(address first, address second) internal pure returns (bool better) {
+        return score(first) > score(second);
+    }
+
+    /// @notice Scores an address based on its vanity
+    /// @param addr The address to score
+    /// @return calculatedScore The vanity score of the address
+    function score(address addr) internal pure returns (uint256 calculatedScore) {
+        //  Scoring rules:
+        //     Requirement: The first nonzero nibble must be 4
+        //     10 points for every leading 0 nibble
+        //     40 points if the first 4 is followed by 3 more 4s
+        //     20 points if the first nibble after the 4 4s is NOT a 4
+        //     20 points if the last 4 nibbles are 4s
+        //     1 point for every 4
+
+        // convert the address to bytes for easier parsing
+        bytes20 addrBytes = bytes20(addr);
+
+        // 10 points per leading zero nibble
+        uint256 leadingZeroCount = getLeadingNibbleCount(addrBytes, 0, 0);
+        calculatedScore += (leadingZeroCount * 10);
+
+        // special handling for 4s immediately after leading 0s
+        uint256 leadingFourCount = getLeadingNibbleCount(addrBytes, leadingZeroCount, 4);
+        // If the first nonzero nibble is not 4, return 0
+        if (leadingFourCount == 0) {
+            return 0;
+        } else if (leadingFourCount == 4) {
+            // 60 points if exactly 4 4s
+            calculatedScore += 60;
+        } else if (leadingFourCount > 4) {
+            // 40 points if more than 4 4s
+            calculatedScore += 40;
```

Comment:
These are not reflected in the scoring rules comment above:
1. the rules say 4 4s score 40 points, but 60 points are in fact scored;
2. the rules don't mention that 4+ 4s are scored less (40) than exactly 4 4s (60).

Reply:
Figured it out, it's implemented as per the rules.